src/libs/common/types/TextDocument.ts

```diff
@@ -0,0 +1,84 @@
+import type { Position, Range, TextLine } from "@cursorless/common";
+import type { URI } from "vscode-uri";
+
+export interface TextDocument {
+  /**
+   * The associated uri for this document.
+   *
+   * *Note* that most documents use the `file`-scheme, which means they are files on disk. However, **not** all documents are
+   * saved on disk and therefore the `scheme` must be checked before trying to access the underlying file or siblings on disk.
+   */
+  readonly uri: URI;
+
+  /**
+   * The identifier of the language associated with this document.
+   */
+  readonly languageId: string;
+
+  /**
+   * The version number of this document (it will strictly increase after each
+   * change, including undo/redo).
+   */
+  readonly version: number;
+
+  /**
+   * The number of lines in this document.
+   */
+  readonly lineCount: number;
+
+  /**
+   * The range of the text document.
+   */
+  readonly range: Range;
+
+  /**
+   * Returns a text line denoted by the line number. Note
+   * that the returned object is *not* live and changes to the
+   * document are not reflected.
+   *
+   * @param line A line number in [0, lineCount).
+   * @return A {@link TextLine line}.
+   */
+  lineAt(line: number): TextLine;
+
+  /**
+   * Returns a text line denoted by the position. Note
+   * that the returned object is *not* live and changes to the
+   * document are not reflected.
+   *
+   * The position will be {@link TextDocument.validatePosition adjusted}.
```

Comment:
This link will be dead, right?  Do we want to reflect this function as well so that we have something to link to?

Reply:
I don't think so. Of course we should not return lines and text that are actually outside of the document. 
I have updated the comment for now.